### PR TITLE
Cancelable ClosureTask

### PR DESCRIPTION
**Pull request by Frago9876543210**
## Introduction
Adds the ability to cancel ClosureTask

## Backwards compatibility
Yes

## Tests
```php
$this->getScheduler()->scheduleDelayedTask(new ClosureTask(function(int $currentTick) : void{
	$this->getLogger()->info("Hello / currentTick = $currentTick");
}), 1); //singe call

//for example, now it will allow make a simple timer
$ticks = 5;
$this->getScheduler()->scheduleRepeatingTask(new ClosureTask(function(int $_) use (&$ticks) : bool{
	if($v = $ticks-- !== 0){
		$this->getLogger()->info("*ticking*");
	}else{
		$this->getLogger()->warning("I'm cancelled!");
	}
	return $v; //if false, it will be cancelled
}, true), 20);
```

### CODE REVIEW

File: src/pocketmine/scheduler/ClosureTask.php [hunk 1]
```diff
@@ -40,20 +40,30 @@ class ClosureTask extends Task{
...
 
 	public function getName() : string{
 		return Utils::getNiceClosureName($this->closure);
 	}
 
 	public function onRun(int $currentTick){
-		($this->closure)($currentTick);
+		if(($this->closure)($currentTick) === false && $this->cancel !== null){
+			$this->getHandler()->cancel();
+			($this->cancel)();
```
Comment (dktapps) [inline src/pocketmine/scheduler/ClosureTask.php]: This can be called from inside `onCancel()` in line with standard tasks.

Comment (Frago9876543210) [reply src/pocketmine/scheduler/ClosureTask.php]: Okey, I'll move it tommorow. That about is_null?

Comment (dktapps) [reply src/pocketmine/scheduler/ClosureTask.php]: Prefer comparison to `null` instead of that.

File: src/pocketmine/scheduler/ClosureTask.php [hunk 2]
```diff
@@ -40,20 +40,33 @@ class ClosureTask extends Task{
...
 	 */
-	public function __construct(\Closure $closure){
-		Utils::validateCallableSignature(function(int $currentTick) : void{}, $closure);
+	public function __construct(\Closure $closure, ?\Closure $cancel = null){
+		Utils::validateCallableSignature(($nullable = $cancel === null) ? function(int $currentTick) : void{} : function(int $currentTick) : bool{ return true; }, $closure);
+		if(!$nullable){
+			Utils::validateCallableSignature(function() : void{}, $cancel);
+		}
 		$this->closure = $closure;
+		$this->cancel = $cancel;
```
Comment (dktapps) [inline src/pocketmine/scheduler/ClosureTask.php]: I've been thinking about this callback and I don't think it makes a whole lot of sense. A self-cancelling task is capable of running actions directly when it does so.

**Review by dktapps — COMMENTED** [review]
I wonder if this logic might be better suited to a separate class instead of butchering `ClosureTask`. I'm not comfortable with the variable signature requirement for the `run` closure.

File: src/pocketmine/scheduler/ClosureTask.php (hunk 2 above)
Comment (Frago9876543210) [reply src/pocketmine/scheduler/ClosureTask.php]: Do you mean that this should be transferred to another class to not to create confusion with the returned type?

Comment (dktapps) [reply src/pocketmine/scheduler/ClosureTask.php]: I mean that this `cancel` closure doesn't make sense.

**Comment by dktapps** [comment]
Obsoleted by 808259dccd650d41fad9c9adfdf5d936a0a7e292.

**Pull request closed without merging**